synapse/membrane.py

```diff
@@ -0,0 +1,166 @@
+# FIXME support ival splices
+# FIXME implement roles/groups/users/perms
+
+import copy
+import fnmatch
+import operator
+
+from synapse.eventbus import EventBus
+import synapse.lib.syntax as s_syntax
+
+class Membrane(EventBus):
+
+    def __init__(self, src=None, dst=None, rules=None, default=None):
+        '''
+        Filters Cortex splices based on rules and forwards to destination Cortex if rules allow.
+
+        Args:
+            dst: the destination Cortex to forward messages to
+            rules: the filter rules
+        '''
+        EventBus.__init__(self)
+
+        _SPLICE_NAMES = (
+            'node:add',
+            'node:del',
+            'node:prop:set',
+            'node:prop:del',
+            'node:tag:add',
+            'node:tag:del',
+            #'node:ival:set',
+            #'node:ival:del',
+        )
+        _DEFAULT_RULE = [{'query': ''}]
+        _DEFAULT_RULES = {'': {k: copy.deepcopy(_DEFAULT_RULE) for k in _SPLICE_NAMES}}
+
+        self.rules = copy.deepcopy(_DEFAULT_RULES)
+        if rules:
+            for key in rules:
+                for rule in rules[key]:
+                    if self._is_valid_rule(rule):
+                        self.rules[''][key].append(rule)
+
+        self.opers = {
+            'eq': operator.eq,
+            'ne': operator.ne,
+            'le': operator.le,
+            'ge': operator.ge,
+            'lt': operator.lt,
+            'gt': operator.gt,
+
+            'has': operator.eq,
+            'tag': fnmatch.fnmatch,
+        }
+
+        self.default = False
+        if default in (True, False):
+            self.default = default
+
+        self.src = src
+        if not self.src:
+            self.src = self
+
+        self.dst = dst
+        if not self.dst:
+            self.dst = self
+
+        self.src.on('splice', self.filter)
+
+    def filter(self, mesg):
+        '''
+        Runs the filter on a given splice message.
+
+        Args:
+            mesg: a synapse splice message
+
+        Returns:
+            result: boolean
+        '''
+        if not(isinstance(mesg, tuple) and len(mesg) == 2 and mesg[0] == 'splice' and isinstance(mesg[1], dict)):
+            print('invalid message: %s' % mesg)
+            return
+
+        result = self._eval_rules(mesg[1])
+        if result is None:
+            result = self.default
+
+        if result:
+            self.dst.splice(mesg)
```

Comment:
should I just the splice pump? would probably work when sending to a core, but not when sending to another membrane.